app/service/rest_svc.py

```diff
@@ -168,11 +168,12 @@ async def get_potential_links(self, op_id, paw=None):
             return []
         agents = await self.get_service('data_svc').locate('agents', match=dict(paw=paw)) if paw else operation.agents
         potential_abilities = await self._build_potential_abilities(operation)
-        links = await self._build_potential_links(operation, agents, potential_abilities)
-        return dict(links=[l.display for l in links])
+        operation.potential_links = await self._build_potential_links(operation, agents, potential_abilities)
+        return dict(links=[l.display for l in operation.potential_links])
 
     async def apply_potential_link(self, link):
-        operation = (await self.get_service('data_svc').locate('operations', match=dict(id=link.operation)))[0]
+        all_operations = await self.get_service('data_svc').locate('operations')
+        operation = next((o for o in all_operations if any(lnk.id == link.id for lnk in o.potential_links)), None)
```

Comment:
This is the second time you've used this exact sequence of code.  Recommend you add a helper function to the `rest_svc` (and maybe a helper function to `c_operation`) to handle this logic.  Idea being you could have a single function like `operation.check_linkid_exists_in_op(link_id)` and it returns True/False (or op_id), then the `rest_svc` helper function would handle 
```Python 
return [o for o in (await self.get_service('data_svc').locate('operations', match=dict(status='running')) if o.check_linkid_exists_in_op(link_id)]
```

Reply:
Good point, since this would be called by `rest_svc` and `contact_svc`, I'm thinking it wouldn't belong in `rest_svc`. Maybe a better placement would be in `app_svc` or changing the `find_link` function that's already in `app_svc`?

Reply:
That would make sense to me.  Thoughts @wbooth ?